// @ledgerhq/wallet-api-core@0.8.0/lib/JSONRPC/helpers.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createRpcResponse = exports.createRpcRequest = exports.parseRPCCall = void 0;
const zod_1 = require("zod");
const RPCError_1 = require("./RPCError");
const types_1 = require("./types");
const validation_1 = require("./validation");
function parseRPCCall(data) {
    try {
        const parsedData = JSON.parse(data);
        const rpcCall = validation_1.schemaRPCCall.parse(parsedData);
        return rpcCall;
    }
    catch (error) {
        if (error instanceof SyntaxError) {
            throw new RPCError_1.RpcError({
                code: types_1.RpcErrorCode.PARSE_ERROR,
                message: "parse error",
            });
        }
        if (error instanceof zod_1.z.ZodError) {
            throw new RPCError_1.RpcError({
                code: types_1.RpcErrorCode.INVALID_REQUEST,
                message: "invalid request",
            });
        }
        throw error;
    }
}
exports.parseRPCCall = parseRPCCall;
function createRpcRequest(params) {
    return {
        jsonrpc: "2.0",
        ...params,
    };
}
exports.createRpcRequest = createRpcRequest;
function createRpcResponse(params) {
    return {
        jsonrpc: "2.0",
        ...params,
    };
}
exports.createRpcResponse = createRpcResponse;
