// @ledgerhq/wallet-api-core@0.9.2/lib/errors/creators.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createUnknownError = exports.createAccountNotFound = exports.createCurrencyNotFound = exports.createPermissionDenied = exports.createNotImplementedByWallet = void 0;
function createNotImplementedByWallet(methodId) {
    return {
        code: "NOT_IMPLEMENTED_BY_WALLET",
        message: "called method is not implemented in the current wallet",
        data: {
            methodId,
        },
    };
}
exports.createNotImplementedByWallet = createNotImplementedByWallet;
function createPermissionDenied(methodId) {
    return {
        code: "PERMISSION_DENIED",
        message: "you don't have the permission to call this method",
        data: {
            methodId,
        },
    };
}
exports.createPermissionDenied = createPermissionDenied;
function createCurrencyNotFound(currencyId) {
    return {
        code: "CURRENCY_NOT_FOUND",
        message: "targeted currency is not supported in the current wallet",
        data: {
            currencyId,
        },
    };
}
exports.createCurrencyNotFound = createCurrencyNotFound;
function createAccountNotFound(accountId) {
    return {
        code: "ACCOUNT_NOT_FOUND",
        message: "targeted account was not found in the current wallet",
        data: {
            accountId,
        },
    };
}
exports.createAccountNotFound = createAccountNotFound;
function createUnknownError(error) {
    return {
        code: "UNKNOWN_ERROR",
        message: "an unhandled error was thrown",
        data: error,
    };
}
exports.createUnknownError = createUnknownError;
