// @ledgerhq/wallet-api-client@0.10.2/lib-es/WalletAPIClient.js
// TODO: reimplement all methods
import { deserializeAccount, Logger, RpcError, RpcErrorCode, RpcNode, schemaAccountList, schemaAccountReceive, schemaAccountRequest, schemaCurrencyList, schemaDeviceTransport, schemaMessageSign, schemaTransactionSign, schemaTransactionSignAndBroadcast, schemaWalletCapabilities, serializeTransaction, } from "@ledgerhq/wallet-api-core";
import { TransportWalletAPI } from "./TransportWalletAPI";
const defaultLogger = new Logger("LL-PlatformSDK");
// temporary
const requestHandlers = {
    "event.account.updated": async (_request) => {
        console.log("accounts updated !");
    },
};
/**
 * WalletAPI Client which rely on WindowMessage communication
 */
export class WalletAPIClient extends RpcNode {
    logger;
    constructor(transport, logger = defaultLogger) {
        super(transport, requestHandlers);
        this.logger = logger;
    }
    onRequest(request) {
        this.logger.log(request.method);
        const handler = this.requestHandlers[request.method];
        if (!handler) {
            throw new RpcError({
                code: RpcErrorCode.METHOD_NOT_FOUND,
                message: "method not found",
            });
        }
        return handler(request);
    }
    /**
     * Let the user sign a transaction that won't be broadcasted by the connected wallet
     * @param accountId - id of the account
     * @param transaction - The transaction object in the currency family-specific format
     * @param options - Extra parameters
     *
     * @returns The raw signed transaction
     * @throws {@link RpcError} if an error occured on server side
     */
    async signTransaction(accountId, transaction, options) {
        const transactionSignResult = await this.request("transaction.sign", {
            accountId,
            rawTransaction: serializeTransaction(transaction),
            options,
        });
        const safeResults = schemaTransactionSign.result.parse(transactionSignResult);
        return Buffer.from(safeResults.signedTransactionHex, "hex");
    }
    /**
     * Let the user sign and broadcast a transaction
     * @param accountId - id of the account
     * @param transaction - The transaction object in the currency family-specific format
     * @param options - Extra parameters
     *
     * @returns The transaction hash
     * @throws {@link RpcError} if an error occured on server side
     */
    async signTransactionAndBroadcast(accountId, transaction, options) {
        const transactionSignResult = await this.request("transaction.signAndBroadcast", {
            accountId,
            rawTransaction: serializeTransaction(transaction),
            options,
        });
        const safeResults = schemaTransactionSignAndBroadcast.result.parse(transactionSignResult);
        return safeResults.transactionHash;
    }
    /**
     * Let the user sign the provided message.
     * In Ethereum context, this is an [EIP-191 message](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-191.md)
     * or an [EIP-712 message](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-712.md)
     * @param accountId - Ledger Live id of the account
     * @param message - Message the user should sign
     *
     * @returns Message signed
     * @throws {@link RpcError} if an error occured on server side
     */
    async signMessage(accountId, message) {
        const messageSignResult = await this.request("message.sign", {
            accountId,
            hexMessage: message.toString("hex"),
        });
        const safeResults = schemaMessageSign.result.parse(messageSignResult);
        return Buffer.from(safeResults.hexSignedMessage, "hex");
    }
    /**
     * List accounts added by user on the connected wallet
     *
     * @param params - Filters for currencies
     *
     * @returns The list of accounts on the connected wallet
     * @throws {@link RpcError} if an error occured on server side
     */
    async listAccounts(params) {
        const listAccountsResult = await this.request("account.list", {
            currencyIds: params?.currencyIds,
        });
        const safeResults = schemaAccountList.result.parse(listAccountsResult);
        return safeResults.rawAccounts.map(deserializeAccount);
    }
    /**
     * Ask the connected wallet for an account matching a specific set of critterias.
     *
     * @param params - Parameters of the request.
     *
     * @returns The account selected by the user
     * @throws {@link RpcError} if an error occured on server side
     */
    async requestAccount(params) {
        const requestAccountsResult = await this.request("account.request", {
            currencyIds: params?.currencyIds,
        });
        const safeResults = schemaAccountRequest.result.parse(requestAccountsResult);
        return deserializeAccount(safeResults.rawAccount);
    }
    /**
     * Let user verify it's account address on his device through Ledger Live
     *
     * @param accountId - id of the account
     *
     * @returns The verified address or an error message if the verification doesn't succeed
     */
    async receive(accountId) {
        const receiveAccountsResult = await this.request("account.receive", {
            accountId,
        });
        const safeResults = schemaAccountReceive.result.parse(receiveAccountsResult);
        return safeResults.address;
    }
    /**
     * List cryptocurrencies supported by the connected wallet, providing filters by name or ticker
     *
     * @param params - Filters for currencies
     *
     * @returns The list of corresponding cryptocurrencies
     * @throws {@link RpcError} if an error occured on server side
     *
     * @beta Filtering not yet implemented
     */
    async listCurrencies(params) {
        const listCurrenciesResult = await this.request("currency.list", {
            currencyIds: params?.currencyIds,
        });
        const safeResults = schemaCurrencyList.result.parse(listCurrenciesResult);
        return safeResults.currencies;
    }
    /**
     * Open low-level transport in the connected wallet
     *
     * @param params - Params for the transport
     *
     * @returns An instance of Transport compatible with @ledgerhq/hw-transport
     * @throws {@link RpcError} if an error occured on server side
     */
    async deviceTransport(params) {
        const deviceTransportResult = await this.request("device.transport", params);
        const safeResults = schemaDeviceTransport.result.parse(deviceTransportResult);
        return TransportWalletAPI.open({
            walletApi: this,
            transportId: safeResults.transportId,
        });
    }
    /**
     * List the wallet's implemented methodIds
     *
     * @returns The list of implemented method ids
     * @throws {@link RpcError} if an error occured on server side
     *
     * @beta Filtering not yet implemented
     */
    async capabilities() {
        const walletCapabilitiesResult = await this.request("wallet.capabilities", {});
        const safeResults = schemaWalletCapabilities.result.parse(walletCapabilitiesResult);
        return safeResults.methodIds;
    }
}
