// @ledgerhq/wallet-api-core@0.9.2/lib/currencies/validation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.schemaCurrency = exports.schemaERC20TokenCurrency = exports.schemaTokenCurrency = exports.schemaCryptoCurrency = exports.schemaBaseCurrency = exports.schemaTokenStandard = exports.schemaCurrencyType = void 0;
const zod_1 = require("zod");
const common_1 = require("../families/common");
exports.schemaCurrencyType = zod_1.z.enum(["CryptoCurrency", "TokenCurrency"]);
exports.schemaTokenStandard = zod_1.z.enum(["ERC20"]);
exports.schemaBaseCurrency = zod_1.z.object({
    color: zod_1.z.string(),
    ticker: zod_1.z.string(),
    id: zod_1.z.string(),
    name: zod_1.z.string(),
    decimals: zod_1.z.number(),
});
exports.schemaCryptoCurrency = exports.schemaBaseCurrency.extend({
    type: zod_1.z.literal(exports.schemaCurrencyType.enum.CryptoCurrency),
    family: common_1.schemaFamilies,
});
exports.schemaTokenCurrency = exports.schemaBaseCurrency.extend({
    type: zod_1.z.literal(exports.schemaCurrencyType.enum.TokenCurrency),
    parent: zod_1.z.string(),
});
exports.schemaERC20TokenCurrency = exports.schemaTokenCurrency.extend({
    standard: exports.schemaTokenStandard,
    contract: zod_1.z.string(),
});
exports.schemaCurrency = zod_1.z.discriminatedUnion("type", [
    exports.schemaCryptoCurrency,
    exports.schemaERC20TokenCurrency,
]);
