// @ledgerhq/wallet-api-core@0.8.0/lib/families/ethereum/serializer.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.deserializeEthereumTransaction = exports.serializeEthereumTransaction = void 0;
const bignumber_js_1 = __importDefault(require("bignumber.js"));
function serializeEthereumTransaction({ family, amount, recipient, nonce, data, gasPrice, gasLimit, }) {
    return {
        family,
        amount: amount.toString(),
        recipient,
        nonce,
        data: data ? data.toString("hex") : undefined,
        gasPrice: gasPrice ? gasPrice.toString() : undefined,
        gasLimit: gasLimit ? gasLimit.toString() : undefined,
    };
}
exports.serializeEthereumTransaction = serializeEthereumTransaction;
function deserializeEthereumTransaction({ family, amount, recipient, nonce, data, gasPrice, gasLimit, }) {
    return {
        family,
        amount: new bignumber_js_1.default(amount),
        recipient,
        nonce,
        data: data ? Buffer.from(data, "hex") : undefined,
        gasPrice: gasPrice ? new bignumber_js_1.default(gasPrice) : undefined,
        gasLimit: gasLimit ? new bignumber_js_1.default(gasLimit) : undefined,
    };
}
exports.deserializeEthereumTransaction = deserializeEthereumTransaction;
